File: src/01_data_structures/08_disjoint_set/implementations/go/disjoint_set_test.go
package disjoint_set

import (
	"reflect"
	"testing"
)

func TestAdjacencyListInstantiation(t *testing.T) {
	disjointSet := NewDisjointSet(3)
	expected := &DisjointSet{representatives: []int{0, 1, 2}, ranks: []int{0, 0, 0}}

	if !reflect.DeepEqual(disjointSet, expected) {
		t.Errorf("Expected %v, got %v", expected, disjointSet)
	}
}

func TestFindsItemInSet(t *testing.T) {
	disjointSet := NewDisjointSet(3)
	representative, err := disjointSet.Find(1)
	if err == nil {
		expected := 1

		if !reflect.DeepEqual(representative, expected) {
			t.Errorf("Expected %v, got %v", representative, expected)
		}
	}
}

func TestErrorsIfIndexOutOfBounds(t *testing.T) {
	disjointSet := NewDisjointSet(3)
	rep, err := disjointSet.Find(2)

	if err != nil {
		t.Errorf("Unexpected error: %v", err)
	}

	if rep != 2 {
		t.Errorf("Expected representative 2, got %v", rep)
	}
}

func TestUnionsGroupsOfGivenRepresentatives(t *testing.T) {
	disjointSet, _ := NewDisjointSet(3).Union(1, 2)

	expected := &DisjointSet{representatives: []int{0, 1, 1}, ranks: []int{0, 1, 0}}

	if !reflect.DeepEqual(disjointSet, expected) {
		t.Errorf("Expected %v, got %v", expected, disjointSet)
	}
}

func TestTwoGroupsConnected(t *testing.T) {
	disjointSet, _ := NewDisjointSet(3).Union(1, 2)
	connected, err := disjointSet.Connected(1, 2)

	if err != nil {
		t.Errorf("Unexpected error: %v", err)
	}

	if !reflect.DeepEqual(connected, true) {
		t.Errorf("Expected %v, got %v", connected, true)
	}
}

func TestConnectedReturnsErrorIfOutOfBound(t *testing.T) {
	disjointSet, _ := NewDisjointSet(3).Union(1, 2)
	connected, err := disjointSet.Connected(1, 4)

	if err == nil {
		t.Errorf("Unexpected error: %v", err)
	}

	if !reflect.DeepEqual(connected, false) {
		t.Errorf("Expected %v, got %v", connected, false)
	}
}

File: src/01_data_structures/08_disjoint_set/implementations/rs/disjoint_set_test.rs
#[cfg(test)]
mod tests {
  use super::super::disjoint_set::{DisjointSet, OUT_OF_BOUNDS_ERROR};

  #[test]
  fn disjoint_set_is_instantiated() {
    let disjoint_set = DisjointSet::new(3);

    let expected = DisjointSet {
      representatives: vec![0, 1, 2],
      ranks: vec![0, 0, 0],
    };
    assert_eq!(disjoint_set, expected);
  }

  #[test]
  fn finds_item_in_set() {
    let mut disjoint_set = DisjointSet::new(3);

    let expected = disjoint_set.find(1);

    assert_eq!(expected, Ok(1));
  }

  #[test]
  fn trows_if_item_out_of_bounds() {
    let mut disjoint_set = DisjointSet::new(3);

    let expected = disjoint_set.find(4);

    assert_eq!(expected, Err(OUT_OF_BOUNDS_ERROR));
  }

  #[test]
  fn unions_groups_of_given_representatives() {
    let mut disjoint_set = DisjointSet::new(3);
    let result = disjoint_set.union(1, 2).unwrap();

    let expected = &DisjointSet {
      representatives: vec![0, 1, 1],
      ranks: vec![0, 1, 0],
    };

    assert_eq!(result, expected);
  }

  #[test]
  fn union_throws_if_representative_out_of_bound() {
    let mut disjoint_set = DisjointSet::new(3);
    let result = disjoint_set.union(1, 4);

    let expected = Err(OUT_OF_BOUNDS_ERROR);

    assert_eq!(result, expected);
  }
}
